lib/kitchen_sink/list/index_by.ex

```diff
@@ -0,0 +1,20 @@
+defmodule KitchenSink.List.IndexBy do
+  @doc """
+  creates and indexed list from a list of maps using a key_function or
+  a path. if a path is given then the access module will be used.
+
+  This function is very similar to group_by, however the values will
+  not be grouped. there should be a 1-to-1 mapping for the key_fun to
+  item in the list.  Do not use this if it's possible that you could
```

Comment:
> Do not use this if it's possible that you could have a different mapping ratio

I the case of having a different mapping ratio, how about we raise instead, with a message that explains the decision/limitation.